parts/linux/cloud-init/artifacts/cse_config.sh

```diff
@@ -744,4 +744,80 @@ EOF
     fi
 }
 
+getPrimaryNicIP() {
+    local sleepTime=1
+    local maxRetries=10
+    local i=0
+    local ip=""
+
+    while [[ $i -lt $maxRetries ]]; do
+        ip=$(curl -sSL -H "Metadata: true" "http://169.254.169.254/metadata/instance/network/interface?api-version=2021-02-01" | jq -r '.[].ipv4.ipAddress[0].privateIpAddress')
+        if [[ -n "$ip" && $? -eq 0 ]]; then
+            break
+        fi
+        sleep $sleepTime
+        i=$((i+1))
+    done
+    echo "$ip"
+}
+
+# $1 - boolean, if true, insert the iptables rule to mangle table; otherwise insert the rule to filter table. Default value is false.
+ensureIMDSRestrictionRule() {
+    local primaryNicIP=$(getPrimaryNicIP)
+    if [[ -z "$primaryNicIP" ]]; then
+        echo "Primary NIC IP not found"
+        exit $ERR_PRIMARY_NIC_IP_NOT_FOUND
+    fi
+    echo "Primary NIC IP: $primaryNicIP"
+
+    local insertRuleToMangleTable=${1:-false}
+    if [[ $insertRuleToMangleTable == true ]]; then
+        echo "Before inserting IMDS restriction rule to mangle table, checking whether the rule already exists..."
+        iptables -t mangle -S | grep -- '-d 169.254.169.254/32 -p tcp -m tcp --dport 80 -m comment --comment "AKS managed: added by AgentBaker esnureIMDSRestriction for IMDS restriction feature" -j DROP'
+        if [[ $? -eq 0 ]]; then
+            echo "IMDS restriction rule already exists in mangle table, returning..."
+            return
+        fi
+        echo "Inserting IMDS restriction rule to mangle table..."
+        iptables -t mangle -I FORWARD 1 ! -s "$primaryNicIP" -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -m comment --comment "AKS managed: added by AgentBaker esnureIMDSRestriction for IMDS restriction feature" -j DROP || exit $ERR_MANIPULATE_IPTABLES
```

Comment:
lets use separate exit codes for each failure here, e.g. `ERR_INSERT_IMDS_RESTRICTION_RULE_INTO_MANGLE`, `ERR_INSERT_IMDS_RESTRICTION_RULE_INTO_FILTER` - same goes for deletion as well

Reply:
Makes great sense, updated.